### ISSUE TITLE
PPC support?

### ISSUE THREAD

**Issue by ZakDanger** [unicorn-engine/unicorn#348]
According to the webpage for unicorn-engine, it supports PPC.
However the readme in this source code as well as the sourcecode itself doesn't appear to support PPC. I noticed there is no "unicorn/ppc.h" file that includes the PPC registers like there is for the other supported platforms. There is an entry in the architecture enums for "UC_ARCH_PPC" though.

So just wondering whats the official word on PPC support?
What needs to be added to support it besides the ppc.h file?

**Comment by aquynh** [comment]
yes PPC support is in plan, but i have no time for that yet.

if somebody can step up for this work, please go ahead.

**Comment by aquynh** [comment]
i removed PPC from the homepage, thanks.

**Comment by zachriggle** [comment]
Any updates for PowerPC?

**Comment by aquynh** [comment]
No, this involves lots of work

**Comment by invano** [comment]
Do you think PPC and PPC64 support is still in plan? Wondering if it's going to happen soonish or it's about making it from scratch.

**Comment by aquynh** [comment]
Quite a lot of people want this, but i have no time for that yet.

**Comment by maximumspatium** [comment]
@aquynh #461 states that Unicorn is based on Qemu 2.2.1. Does it still hold true today?

Did I get it right that adding PowerPC support to Unicorn means pulling `target-ppc` code from Qemu 2.2.1 and adopting it to Unicorn's requirements?

**Comment by Vociferix** [comment]
I just want to bump this issue. Like @maximumspatium I might be interested in taking on some of this work if there were some amount of documentation on what all needs to be done to port an arch from Qemu to Unicorn.

**Comment by mePy2** [comment]
+1

**Comment by aquynh** [comment]
sorry that there is no documentation, except source code.

you can diff the current source with Qemu 2.1.2, to figure out the
necessary modification for a new arch to be ported.

PPC is one of the most complicated arch to port.

**Comment by simigo79** [comment]
Hi, I've done (hopefully) most of the work. At this point I have 32 bit ppc initializing, setting up emulation, starting emulation, fetching an instruction then producing segfault somewhere in tcg.
The code is based on this git cloned on Oct 10 2019 and qemu 2.2.1
As far as I understand the code can't be uploaded on github now because it is a work and progress and have copyright notices, etc not updated in the changed files. Anyway if someone feels he can really help please contact me. Otherwise just wait. :-)

**Comment by aquynh** [comment]
nice work! i can wait ;-)

**Comment by simigo79** [comment]
Good news, everyone! I have a very basic test working.

**Comment by simigo79** [comment]
Forked: https://github.com/simigo79/unicorn-ppc

**Comment by aquynh** [comment]
Amazing work! Is this ready to use, or you are still working on it?


Do you mind making pull req?

**Comment by simigo79** [comment]
Well, I've made only few tests so far: basic arithmetic, and strlen(), no stack-related tests, no tests for resource leaks on multiple runs. Also it is limited to e500v2 and built only in ubuntu/amd64/gcc environment. So this is surely work in progress. 
However the changes to the common part of unicorn are few and easy to review and most likely they are not going to introduce any problem with non-ppc platforms. So I created a pull request, if you accept the changes you are going to have an otherwise stable engine with experimental unstable ppc support.
